File: gcc/testsuite/gfortran.dg/ISO_Fortran_binding_10.c
/* Test the fix of PR89843.  */

/* Contributed by Reinhold Bader  <Bader@lrz.de> */

#include "../../../libgfortran/ISO_Fortran_binding.h"
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

void sa(CFI_cdesc_t *, int, int *);

void si(CFI_cdesc_t *this, int flag, int *status)
{
  int value, sum;
  bool err;
  CFI_CDESC_T(1) that;
  CFI_index_t lb[] = { 0, 0 };
  CFI_index_t ub[] = { 4, 0 };
  CFI_index_t st[] = { 2, 0 };
  int chksum[] = { 9, 36, 38 };

  if (flag == 1)
    {
      lb[0] = 0; lb[1] = 2;
      ub[0] = 2; ub[1] = 2;
      st[0] = 1; st[1] = 0;
    }
  else if (flag == 2)
    {
      lb[0] = 1; lb[1] = 0;
      ub[0] = 1; ub[1] = 3;
      st[0] = 0; st[1] = 1;
    }

  CFI_establish((CFI_cdesc_t *) &that, NULL, CFI_attribute_other,
		CFI_type_float, 0, 1, NULL);

  *status = CFI_section((CFI_cdesc_t *) &that, this, lb, ub, st);

  if (*status != CFI_SUCCESS)
    {
      printf("FAIL C: status is %i\n",status);
      return;
    }

  value = CFI_is_contiguous((CFI_cdesc_t *) &that);
  err = ((flag == 0 && value != 0)
	 || (flag == 1 && value != 1)
	 || (flag == 2 && value != 0));

  if (err)
    {
      printf("FAIL C: contiguity for flag value %i - is %i\n", flag, value);
      *status = 10;
      return;
    }

  sum = 0;
  for (int i = 0; i < that.dim[0].extent; i++)
    {
      CFI_index_t idx[] = {i};
      sum += (int)(*(float *)CFI_address ((CFI_cdesc_t *)&that, idx));
    }

  if (sum != chksum[flag])
    {
      printf ("FAIL C: check sum = %d(%d)\n", sum, chksum[flag]);
      *status = 11;
      return;
    }

    sa((CFI_cdesc_t *) &that, flag, status);
}

File: gcc/testsuite/gfortran.dg/ISO_Fortran_binding_10.f90
! { dg-do run { target c99_runtime } }
! { dg-additional-sources ISO_Fortran_binding_10.c }
!
! Test the fix of PR89843.
!
! Contributed by Reinhold Bader  <Bader@lrz.de>
!
module mod_section_01
  use, intrinsic :: iso_c_binding
  implicit none
  interface
     subroutine si(this, flag, status) bind(c)
       import :: c_float, c_int
       real(c_float) :: this(:,:)
       integer(c_int), value :: flag
       integer(c_int) :: status
     end subroutine si
  end interface
contains
  subroutine sa(this, flag, status) bind(c)
    real(c_float) :: this(:)
    integer(c_int), value :: flag
    integer(c_int) :: status

    status = 0

    select case (flag)
    case (0)
       if (is_contiguous(this)) then
          write(*,*) 'FAIL 1:'
          status = status + 1
       end if
       if (size(this,1) /= 3) then
          write(*,*) 'FAIL 2:',size(this)
          status = status + 1
          goto 10
       end if
       if (maxval(abs(this - [ 1.0, 3.0, 5.0 ])) > 1.0e-6) then
          write(*,*) 'FAIL 3:',abs(this)
          status = status + 1
       end if
  10   continue
   case (1)
      if (size(this,1) /= 3) then
          write(*,*) 'FAIL 4:',size(this)
          status = status + 1
          goto 20
       end if
       if (maxval(abs(this - [ 11.0, 12.0, 13.0 ])) > 1.0e-6) then
          write(*,*) 'FAIL 5:',this
          status = status + 1
       end if
  20   continue
   case (2)
      if (size(this,1) /= 4) then
          write(*,*) 'FAIL 6:',size(this)
          status = status + 1
          goto 30
       end if
      if (maxval(abs(this - [ 2.0, 7.0, 12.0, 17.0 ])) > 1.0e-6) then
          write(*,*) 'FAIL 7:',this
          status = status + 1
       end if
  30   continue
    end select

!    if (status == 0) then
!       write(*,*) 'OK'
!    end if
  end subroutine sa
end module mod_section_01

program section_01
  use mod_section_01
  implicit none
  real(c_float) :: v(5,4)
  integer :: i
  integer :: status

  v = reshape( [ (real(i), i = 1, 20) ], [ 5, 4 ] )
  call si(v, 0, status)
  if (status .ne. 0) stop 1

  call sa(v(1:5:2, 1), 0, status)
  if (status .ne. 0) stop 2

  call si(v, 1, status)
  if (status .ne. 0) stop 3

  call sa(v(1:3, 3), 1, status)
  if (status .ne. 0) stop 4

  call si(v, 2, status)
  if (status .ne. 0) stop 5

  call sa(v(2,1:4), 2, status)
  if (status .ne. 0) stop 6

end program section_01
